### PR TITLE
SK-265 initiate a local DB to store messages

### PR DESCRIPTION
**Pull request by Oleksandr1414**

### CODE REVIEW

File: src/store/values/Conversations.js [hunk 1]
```diff
@@ -84,7 +84,13 @@ export const conversations = createSlice({
 
       const { messagesIds, unread_messages_count } =
         state.entities[conversation._id] || {};
-      // messagesIds && delete conversation.messagesIds;
+      messagesIds &&
```
Comment (IgorKhomenko) [inline src/store/values/Conversations.js:87]: can make this code a bit easier to read 
```
if (messagesIds) {
  conversation.messagesIds =  (conversation.messagesIds ?? []).concat(messagesIds)
}
```

File: src/services/messagesService.js [hunk 2]
```diff
@@ -268,8 +285,44 @@ class MessagesService {
...
+    );
+
+    store.dispatch(addMessages(decryptedMessages));
+    store.dispatch(insertChats(conversations));
+  }
 }
 
 const messagesService = new MessagesService();
 
+window.onbeforeunload = messagesService.saveMessagesToLocalStorage;
```
Comment (IgorKhomenko) [inline src/services/messagesService.js:326]: will this event work if user kills the browser?

Comment (Oleksandr1414) [reply src/services/messagesService.js:326]: yes

File: src/components/info/UserProfile.js [hunk 3]
```diff
@@ -46,10 +47,12 @@ export default function UserProfile() {
     try {
       await usersService.logout();
       await encryptionService.clearStoredAccount();
+      await messagesService.clearLocalMessages();
```
Comment (IgorKhomenko) [inline src/components/info/UserProfile.js:50]: we already have all these cleaning inside `userService.logout()`
https://github.com/SAMA-Communications/sama-client/blob/35b68fffd03fa9c2b590aaf9d234fff7fd71cd4c/src/services/usersService.js#L131

so no need to do it here

File: src/services/autoLoginService.js [hunk 4]
```diff
@@ -26,6 +27,7 @@ class AutoLoginService {
     const currentPath = history.location?.hash;
     const handleLoginFailure = () => {
       encryptionService.clearStoredAccount();
+      messagesService.clearLocalMessages();
```
Comment (IgorKhomenko) [inline src/services/autoLoginService.js:30]: What I'm worried is we have this code

```
encryptionService.clearStoredAccount();
messagesService.clearLocalMessages();
```

in 4 or 5 diff places in app. So whenever we want to add another cleaning, we need to add it in all these places.
Maybe we can somehow optimise it.

File: src/components/hub/ChatForm.js [hunk 5]
```diff
@@ -57,6 +58,7 @@ export default function ChatForm() {
     dispatch(setClicked(false));
     dispatch(clearSelectedConversation());
     api.unsubscribeFromUserActivity({});
+    messagesService.clearPrevConvMessagesToLocalLimit(selectedCID);
```
Comment (IgorKhomenko) [inline src/components/hub/ChatForm.js:61]: do we also do this cleaning when switch between convs?

File: src/store/indexedDB.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  markMessagesAsRead(mids) {
+    this.db.messages.bulkUpdate(
+      mids.map((id) => ({ key: id, changes: { status: "read" } }))
+    );
+  }
+
+  async addMessage(message) {
+    const encryptedBody = encryptionService.encrypteDataForLocalStore(
```
Comment (IgorKhomenko) [inline src/store/indexedDB.js:20]: In a case we are in a regular chat (not encrypted), how this method will work? 
For regular chats we just need to store a message, w/o encrypting it

File: src/components/hub/elements/MessagesList.js [hunk 7]
```diff
@@ -56,74 +57,85 @@ export default function MessagesList({ scrollRef }) {
...
-              };
-              return;
-            }
+      }
 
-            const mids = mAttachmentsObject._id;
-            mAttachments[obj.file_id]._id = Array.isArray(mids)
-              ? [arr[i]._id, ...mids]
-              : [arr[i]._id, mids];
+      if (Object.keys(mAttachments).length > 0) {
```
Comment (IgorKhomenko) [inline src/components/hub/elements/MessagesList.js:113]: why we have this code 2 times?

```
  if (Object.keys(mAttachments).length > 0) {
```

File: src/components/hub/elements/MessagesList.js [hunk 8]
```diff
@@ -56,74 +57,85 @@ export default function MessagesList({ scrollRef }) {
...
+      dispatch(
+        upsertChat({
+          _id: selectedCID,
+          messagesIds: [
+            ...new Set([...messagesIds, ...messages.map((el) => el._id)]),
+          ],
+          activated: true,
+        })
+      );
+      const mAttachments = {};
```
Comment (IgorKhomenko) [inline src/components/hub/elements/MessagesList.js:84]: To be honest, this code looks very similar to `messageService.syncData`. Can we reuse/optimise?

File: src/components/hub/ChatForm.js (hunk 5 above)
Comment (Oleksandr1414) [reply src/components/hub/ChatForm.js]: vvv
https://github.com/SAMA-Communications/sama-client/pull/135/files#diff-f13243307d8bf1149546c0e0e90c863503bece4d3d008d43bf5a4caec63556f6R164

File: src/services/garbageCleaningService.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import localforage from "localforage";
+import store from "@store/store";
+import { clearMessageIdsToLocalLimit } from "@store/values/Conversations";
+import { clearMessagesToLocalLimit } from "@store/values/Messages";
+import { setSelectedConversation } from "@store/values/SelectedConversation";
+import { setUserIsLoggedIn } from "@store/values/UserIsLoggedIn";
+import { updateNetworkState } from "@store/values/NetworkState";
+
+class GarbageCleaningService {
+  async #clearAccountData() {
```
Comment (IgorKhomenko) [inline src/services/garbageCleaningService.js:13]: `#clearAccountData()` and `#clearAccountDataSync()` methods look the same

File: src/services/garbageCleaningService.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+  #clearLocalMessageIds(cid) {
+    store.dispatch(clearMessageIdsToLocalLimit(cid));
+  }
+
+  #clearLocalMessages(cid) {
+    store.dispatch(clearMessagesToLocalLimit(cid));
+  }
+
+  async clearConversationMessages(cid) {
```
Comment (IgorKhomenko) [inline src/services/garbageCleaningService.js:29]: do we need separate methods for 
`#clearLocalMessageIds` and `#clearLocalMessageIds` ? Will it be a case where we need to call it separately? Probably not.

so we can optimise

```
async clearConversationMessages(cid) {
  if (!cid) return;
  store.dispatch(clearMessagesToLocalLimit(cid));
  store.dispatch(clearMessageIdsToLocalLimit(cid));
}
```

File: src/services/garbageCleaningService.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  async resetDataOnAuth() {
+    await this.#clearAccountData();
+    await this.clearAllLocalData();
+    store.dispatch(setSelectedConversation({}));
+    store.dispatch(setUserIsLoggedIn(true));
+  }
+
+  async handleLogout() {
+    await api.userLogout();
+    await this.#clearAccountData();
```
Comment (IgorKhomenko) [inline src/services/garbageCleaningService.js:44]: The naming is confusing a bit, e.g.

`clearAccountData` and `clearAllLocalData`

Maybe we can have only one method:

```
 async clearStorage() {
   await Promise.all([localforage.clear(), indexedDB.removeAllMessages()]);
   await encryptionService.clearStoredAccount();
 }
```

File: src/services/messagesService.js [hunk 12]
```diff
@@ -157,82 +161,114 @@ class MessagesService {
...
+              : [msg._id, existingId];
+          }
+        });
+      }
+      return acc;
+    }, {});
+    return attachments;
+  }
+
+  async updateMessages(attachments) {
```
Comment (IgorKhomenko) [inline src/services/messagesService.js:188]: `updateMessages(attachments)` -> `retrieveAttachmentsLinks(attachments) `

File: src/services/encryptionService.js [hunk 13]
```diff
@@ -37,7 +39,7 @@ class EncryptionService {
   async clearStoredAccount() {
     this.#account = null;
     this.#encryptionSessions = {};
-    await localforage.clear();
+    await garbageCleaningService.clearAllLocalData();
```
Comment (IgorKhomenko) [inline src/services/encryptionService.js:42]: this looks like a cycle, e.g. GC service calls EncryptionService which calls GC service etc..

File: src/components/hub/elements/MessagesList.js [hunk 14]
```diff
@@ -56,74 +56,49 @@ export default function MessagesList({ scrollRef }) {
...
-          needToGetMoreMessage.current = false;
-          return;
-        }
-
-        const messagesIds = arr.map((el) => el._id).reverse();
-        needToGetMoreMessage.current = !(
-          messagesIds.length < +process.env.REACT_APP_MESSAGES_COUNT_TO_PRELOAD
-        );
+    const processMessages = (arr) => {
+      indexedDB.insertManyMessages(arr);
```
Comment (IgorKhomenko) [inline src/components/hub/elements/MessagesList.js:60]: let's make `messagesService ` responsible for communicating with `indexedDB `, 
so `MessagesList` (UI component) should not know which DB is used

File: src/components/hub/elements/MessagesList.js [hunk 15]
```diff
@@ -56,74 +55,42 @@ export default function MessagesList({ scrollRef }) {
...
+          _id: selectedCID,
+          messagesIds: [
+            ...new Set([...messagesIds, ...messages.map((el) => el._id)]),
+          ],
+          activated: true,
+        })
+      );
+
+      const mAttachments = messagesService.processAttachments(arr);
+      messagesService.retrieveAttachmentsLinks(mAttachments);
```
Comment (IgorKhomenko) [inline src/components/hub/elements/MessagesList.js:82]: I have the very last comment

This code is duplicated in `messageService.syncData`, hence we can refactor it a bit to reduce code duplication.

One possible way to do so is to move this code directly into  `messageService.retrieveMessages`, e.g.


```
  async retrieveMessages(params) {
    const messagesDB = await indexedDB.getMessages(params);

    if (messagesDB.length >= params.limit) return messagesDB;
    const messagesAPI = await api.messageList(params);
    await indexedDB.insertManyMessages(messagesAPI);
    
   const messagesIds = messages.map((el) => el._id).reverse();

   const jointMessages =  [...messagesDB, ...messagesAPI]; // do we also need to pass here ids from redux?
   
   // save to redux store 
   store.dispatch(addMessages(jointMessages));
   store.dispatch(
      upsertChat({ _id: this.currentChatId, messagesIds, activated: true })
   );

   const mAttachments = messagesService.processAttachments(arr);
   messagesService.retrieveAttachmentsLinks(mAttachments);
  }
```

WDYT?

**Pull request merged**